tests/core/kernel-and-base-single-reboot-failover/task.yaml

```diff
@@ -31,49 +47,87 @@ execute: |
+        fi
+
+        snap refresh --no-wait "$SNAP_ONE" "$SNAP_TWO" > refresh-change-id
         test -n "$(cat refresh-change-id)"
         change_id="$(cat refresh-change-id)"
-        # wait until we observe reboots
-        # shellcheck disable=SC2016
-        retry -n 100 --wait 5 sh -c 'test "$(wc -l < /tmp/mock-shutdown.calls)" -gt "1"'
-        # stop snapd now to avoid snapd waiting for too long and deciding to
-        # error out assuming a rollback across reboot
-        systemctl stop snapd.service snapd.socket
+        
+        # wait for the link tasks to be done
+        retry -n 50 --wait 1 sh -c 'journalctl -b -u snapd | MATCH "Waiting for system reboot"'
 
         # both link snaps should be done now, snapd was stopped, so we cannot
         # use 'snap change' and we need to inspect the state directly (even if
         # snapd was up, it would not respond to API requests as it would be busy
         # retrying auto-connect)
         snap debug state --change "$change_id" /var/lib/snapd/state.json > tasks.state
         # both link snaps are done
-        MATCH ' Wait\s+.*Make snap "pc-kernel" .* available' < tasks.state
-        MATCH ' Done\s+.*Make snap "core18" .* available' < tasks.state
+        MATCH " Wait\s+.*Make snap \"$SNAP_TWO\" .* available" < tasks.state
+        MATCH " Done\s+.*Make snap \"$SNAP_ONE\" .* available" < tasks.state
         # auto-connect of the base is in doing and waiting for reboot
-        MATCH ' Do\s+.*Automatically connect eligible plugs and slots of snap "core18"' < tasks.state
+        MATCH " Do\s+.*Automatically connect eligible plugs and slots of snap \"$SNAP_ONE\"" < tasks.state
         # auto-connect of the kernel is still queued
-        MATCH ' Do\s+.*Automatically connect eligible plugs and slots of snap "pc-kernel"' < tasks.state
-
-        snap debug boot-vars > boot-vars.dump
-        MATCH 'snap_mode=try' < boot-vars.dump
-        MATCH 'snap_try_core=core18_.*.snap' < boot-vars.dump
-        MATCH 'snap_try_kernel=pc-kernel_.*.snap' < boot-vars.dump
-
-        # restore shutdown so that spread can reboot the host
-        tests.cleanup pop
+        MATCH " Do\s+.*Automatically connect eligible plugs and slots of snap \"$SNAP_TWO\"" < tasks.state
+
+        if os.query is-core18; then
+            snap debug boot-vars > boot-vars.dump
+            if [[ "$SNAP_ONE" == *"core"* ]] || [[ "$SNAP_TWO" == *"core"* ]]; then
+                MATCH 'snap_mode=try' < boot-vars.dump
+                MATCH 'snap_try_core=core18_.*.snap' < boot-vars.dump
+            fi
+            if [[ "$SNAP_ONE" == "pc-kernel" ]] || [[ "$SNAP_TWO" == "pc-kernel" ]]; then
+                MATCH 'snap_mode=try' < boot-vars.dump
+                MATCH 'snap_try_kernel=pc-kernel_.*.snap' < boot-vars.dump
+            fi
+        elif os.query is-core20; then
+            if [[ "$SNAP_ONE" == "pc-kernel" ]] || [[ "$SNAP_TWO" == "pc-kernel" ]]; then
```

Comment:
Iirc what would need to be checked in the case of a base reboot for UC20+ is the modeenv file (`try_base`). And still check grubenv for the kernel update.

Reply:
Will add this